lib/hotel_manager.rb

```diff
@@ -0,0 +1,194 @@
+      reservation = Reservation.new(date_range, room: room)
+      
+      self.add_reservation(reservation, type: :individual)
+
+      return reservation
+    end 
+
+
+    #  Find reservations per room withinn specific date range (using overlap?)
+    def reservations_per_room(date_range, room)
+      return @reservations[:individual].filter do |reservation|
+        (reservation.room.id == room.id) && (reservation.date_range.overlap?(date_range))
+      end 
+    end 
+    
+    
+    # Find reservations for the specific date (using include?)
+    def find_reservations(date)
+      return @reservations[:individual].find_all do |reservation|
+          (reservation.date_range.include?(date)) && (reservation.room) 
+      end 
+    end 
+
+    
+    def reserved_room_ids(date_range) 
+      ids = []
+
+      @reservations[:individual].each do |reservation| 
+        if reservation.date_range.overlap?(date_range)
+          ids << reservation.room.id
+        end 
+      end 
+
+      @reservations[:block].each do |reservation|  
+        if reservation.block && reservation.block.date_range.overlap?(date_range)
+          reservation.block.rooms.each do |block_room|
+            ids << block_room.id
+          end 
+        end 
+      end 
+
+      # Need to filter reserved blocks as well
+      @blocks.each do |block|
+        if block.date_range.overlap?(date_range)
+          block.rooms.each do |room|
+            ids << room.id
+          end 
+        end 
+      end 
+
+      return ids
+    end
+
+
+    def available_room_ids(date_range)
+      ids = (1..20).to_a
+      reserved_room_ids = reserved_room_ids(date_range)
+
+      available_room_ids = ids - reserved_room_ids
+
+      return available_room_ids
+    end 
+
+
+     # ====== block =====
+
+    def set_block(date_range, number_of_rooms: 1)
+      # Change test file
+      if (number_of_rooms < 1) || (number_of_rooms > 5)
+        raise ArgumentError, "A block can contain between 1 and 5 rooms (maximum of 5 rooms)"
+      end 
+
+      # Find available rooms for the specific date range
+      available_room_ids = available_room_ids(date_range) 
+      
+      if (available_room_ids.length < number_of_rooms) 
+        raise ArgumentError, "We have only #{available_room_ids.length} rooms between #{date_range.start_date} and #{date_range.end_date} you chose"
+      end 
+
+      rooms = []
+
+      number_of_rooms.times do |i|
+        id = available_room_ids[-1]
+
+        room = find_room_by_id(id)
+        rooms << room
+
+        available_room_ids.delete(id)
+      end 
+
+      block = Hotel::Block.new(date_range, rooms: rooms) 
+
+      @blocks << block
+      return block
+    end 
+
+
+    def available_blocks(date_range, number_of_rooms:)
+      return @blocks.find_all do |block|
+        block.date_range.same?(date_range) && (block.rooms.length == number_of_rooms)
+      end 
+    end 
+
+    
+    def make_block_reservation(date_range, number_of_rooms: 1)
```

Comment:
Per user stories, blocks need to have a name or special id associated with them. I'm not sure I see how this functionality is built in. I think this would just be a matter of adding an attribute to the block class. Or maybe this is accounted for with the reservation id? "These rooms are set aside, and are made available for reservation by certain customers at a discounted rate."

Reply:
Thank you Becca for checking them in detail. 🙏 That was one of my challenges and I felt that my design might be too complicated. Like you guessed, for the latter part ("maybe this is accounted for with the reservation id? "), I made a Block connected with Reservation. For instance, (1) if it's a block, Reservation has a block and the Block has rooms, whereas (2) If it's a regular reservation, Reservation has a room instead of Block. 

<Summary>
AS IS
----------------------------------------------------
Block: Reservation -> Block -> Rooms
Individual: Reservation -> Room
----------------------------------------------------
My question is "Would it be better to separate BlockReservation and Reservation and make them like Polymorphism as written below"? 

TO BE
----------------------------------------------------
BlockReservation -> Rooms 
Reservation -> Rooms
----------------------------------------------------

I really appreciate your feedback! Thank you. 🥰